support re-creating/destroying process groups when some trainers recover after failures

mehta-vikas opened the pull request:
Summary:
group name is used as prefix in the c10d store and without a consistent name process group cannot be initialized.

If process group doesn't have a name (only default process group can have non-default name), it is using the counter as name. We need to decrement the global counter on destruction to allow to consistent value to be generated when we re-create process groups to support trainers recovering from failures.

Test Plan: existing tests passed

Differential Revision: D17594268

pietern commented on torch/distributed/distributed_c10d.py:
```diff
@@ -527,12 +528,21 @@ def destroy_process_group(group=group.WORLD):
         del _pg_names[pg]
         del _pg_group_ranks[pg]
 
-
```
@mehta-vikas Double newline between functions is gone.

mehta-vikas replied:
incorporated comments

pietern reviewed (approved):
Please fix the newline before merging.

The title should also be updated: it talks about non-default process groups, whereas this holds for both the default and non-default process groups.

mehta-vikas committed:
support re-creating/destroying process groups when some trainers recover after failures (#26912)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/26912

group name is used as prefix in the c10d store and without a consistent name process group cannot be initialized.

When process group doesn't have an explicit name (only WORLD (default) process group can have an explicit name), we use global _group_counter to generate the name. We need to reset the counter on destruction to allow consistent value to be generated when we re-create process groups after some trainers recover from failure.

Test Plan: existing tests passed

Reviewed By: mrshenli

Differential Revision: D17594268

fbshipit-source-id: 96dd4a234e7e115d29153df71eda0753ed7770f2
